### PR TITLE
Calculates mgmt port MAC rather than storing

### PR DESCRIPTION
**Pull request by trozet**
Today ovnkube-controller relies on ovnkube-node to annotate the mac address so that it can properly program the OVN mgmt port. In practice with UDN, this means cluster manager creates the node and allocates some subnet info to the node, both the node and the ovnkube-controller side get updates, but ovnkube-controller fails to program the first time because it is waiting for the node side to configure the mac address.

This patch changes the behavior of ovn-kubernetes to calculate the MAC address for the management port from the mgmt port IP address of the first subnet for the network. For backwards compatibility (on default network), it will first attempt to read the node annotation, and if it no mgmt MAC exists, it will derive it from the mgmt IP of the subnet.

Note the DPU code already uses the mgmt IP to derive the MAC.

### CODE REVIEW

**Comment by trozet** [comment]
some unrelated egress firewall flake:

https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/12261395552/job/34208326354?pr=4882

File: go-controller/pkg/ovn/base_network_controller.go [hunk 1]
```diff
@@ -627,10 +627,21 @@ func (bnc *BaseNetworkController) deleteNamespaceLocked(ns string) (*namespaceIn
...
 
 func (bnc *BaseNetworkController) syncNodeManagementPort(node *kapi.Node, switchName, routerName string, hostSubnets []*net.IPNet) ([]net.IP, error) {
-	macAddress, err := util.ParseNodeManagementPortMACAddresses(node, bnc.GetNetworkName())
-	if err != nil {
+	// get mac address from node only for legacy reasons, if it doesn't exist, then calculate it from subnets
+	var macAddress net.HardwareAddr
+	var err error
+	// find suitable MAC address
+	// check node annotation first, to ensure we are not picking a new MAC when one was already configured
+	if macAddress, err = util.ParseNodeManagementPortMACAddresses(node, bnc.GetNetworkName()); err != nil && !util.IsAnnotationNotSetError(err) {
```
Comment (kyrtapz) [inline go-controller/pkg/ovn/base_network_controller.go:635]: When configuring the management port in UDN we always use the IP address as base, so there is no point of looking for annotation for UDN networks?
Do you plan to handle UDNs that have the annotation?

Comment (trozet) [reply go-controller/pkg/ovn/base_network_controller.go:635]: >When configuring the management port in UDN we always use the IP address as base, so there is no point of looking for annotation for UDN networks?

Right.

>Do you plan to handle UDNs that have the annotation?

Originally I had this same check for if the node had the UDN mac annotation we would use it. This was for handling upgrade where UDNs were already programmed. After thinking about it some, I realized UDN is not GA so we don't need to handle this case. We can make sure this gets backported to 4.17z and pretend UDN never used that annotation :)

File: go-controller/pkg/node/gateway_udn.go [hunk 2]
```diff
@@ -401,39 +397,35 @@ func (udng *UserDefinedNetworkGateway) DelNetwork() error {
...
-	}
-	// persist the MAC address so that upon node reboot we get back the same mac address.
-	_, stderr, err = util.RunOVSVsctl("set", "interface", interfaceName,
-		fmt.Sprintf("mac=%s", strings.ReplaceAll(macAddress.String(), ":", "\\:")))
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to persist MAC address %q for %q while plumbing network %s: stderr:%s (%v)",
-			macAddress.String(), interfaceName, udng.GetNetworkName(), stderr, err)
-	}
-
-	// STEP3
```
Comment (kyrtapz) [inline go-controller/pkg/node/gateway_udn.go:433]: NIT: There is a STEP4 comment below

File: go-controller/pkg/ovn/base_network_controller_secondary.go [hunk 3]
```diff
@@ -818,11 +818,17 @@ func (bsnc *BaseSecondaryNetworkController) buildUDNEgressSNAT(localPodSubnets [
...
 		return nil, fmt.Errorf("failed to get networkID for network %q: %v", bsnc.GetNetworkName(), err)
 	}
+	// legacy lookup for mac
 	dstMac, err := util.ParseNodeManagementPortMACAddresses(node, bsnc.GetNetworkName())
-	if err != nil {
+	if err != nil && !util.IsAnnotationNotSetError(err) {
 		return nil, fmt.Errorf("failed to parse mac address annotation for network %q on node %q, err: %w",
 			bsnc.GetNetworkName(), node.Name, err)
 	}
+	if len(dstMac) == 0 && len(localPodSubnets) > 0 {
```
Comment (kyrtapz) [inline go-controller/pkg/ovn/base_network_controller_secondary.go:827]: NIT: localPodSubnets length is already checked at the beginning of the functon

File: go-controller/pkg/ovn/base_network_controller.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/ovn/base_network_controller.go:635]: That shouldn't happen, but I can make a guard checking if this controller is default or not for the annotation check. Does that work for you?

Comment (kyrtapz) [reply go-controller/pkg/ovn/base_network_controller.go:635]: I am ok to keep it as is if this is not a possible issue.

**Review by kyrtapz — DISMISSED** [review]

**Review by kyrtapz — APPROVED** [review]

File: go-controller/pkg/node/gateway_udn.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/node/gateway_udn.go:433]: fixed

**Comment by trozet** [comment]
I missed updating some UDN unit tests which were depending on the annotation. Updating those now in next push.

**Review by kyrtapz — APPROVED** [review]

**Pull request merged**